src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+	fprintf(stderr, "full command line: %s\n", command->str);
+#endif
+
+	while (envp && *envp)
+	{
+		g_array_append_vals(environment, *envp, strlen(*envp) + 1);
+		envp++;
+	}
+	
+	failure = spawn_create_process_with_pipes(command->str, working_directory,
+		envp ? environment->data : NULL, child_pid ? child_pid : &pid,
+		stdin_fd, stdout_fd, stderr_fd);
+
+	g_string_free(command, TRUE);
+	g_array_free(environment, TRUE);
+
+	if (failure)
+	{
+		g_set_error(error, G_SPAWN_ERROR, G_SPAWN_ERROR_FAILED, "%s", failure);
+		g_free(failure);
+		return FALSE;
+	}
+	else if (!child_pid)
+		g_child_watch_add(pid, spawn_close_pid, NULL);
+
+	return TRUE;
+#else  /* G_OS_WIN32 */
+	int cl_argc;
+	char **full_argv;
+	gboolean spawned;
+
+	if (command_line)
+	{
+		int argc = 0;
+		char **cl_argv;
+
+		if (!g_shell_parse_argv(command_line, &cl_argc, &cl_argv, error))
+			return FALSE;
+
+		if (argv)
+			for (argc = 0; argv[argc]; argc++);
+
+		full_argv = g_new(gchar *, cl_argc + argc + 1);
+		memcpy(full_argv, cl_argv, cl_argc * sizeof(gchar *));
+		memcpy(full_argv + cl_argc, argv, argc * sizeof(gchar *));
+		full_argv[cl_argc + argc] = NULL;
+	}
+	else
+		full_argv = argv;
+
+	spawned = g_spawn_async_with_pipes(working_directory, full_argv, envp,
+		G_SPAWN_SEARCH_PATH | (child_pid ? G_SPAWN_DO_NOT_REAP_CHILD : 0), NULL, NULL,
+		child_pid, stdin_fd, stdout_fd, stderr_fd, error);
+
+	if (spawned)
+	{
+		int *fd[3] = { stdin_fd, stdout_fd, stderr_fd };
+		int i;
+
+		for (i = 0; i < 3; i++)
+			if (fd[i])
+				fcntl(*fd[i], F_SETFL, O_NONBLOCK);
+	}
+
+	if (full_argv != argv)
+	{
+		full_argv[cl_argc] = NULL;
+		g_strfreev(full_argv);
+	}
+
+	return spawned;
+#endif  /* G_OS_WIN32 */
+}
+
+
+enum { OUT, ERR, IN };
+
+typedef struct _SpawnSource
+{
+	GSource source;
+	GPollFD std[3];
+	gchar *stdin_text;
+	const char *stdin_pos;
+	size_t stdin_remaining;
+#ifdef G_OS_WIN32
+	DWORD last_send_ticks;
+#endif
+	GIOChannel *ioc[2];
+	void (*io_cb[2])(const gchar *text, gpointer user_data);
+	gpointer io_data[2];
+	GPid pid;
+	void (*exit_cb)(gint status, gpointer user_data);
+	gpointer exit_data;
+	/* interactive */
+	GtkWidget *dialog;
+	const gchar *command;
+} SpawnSource;
+
+
+static void spawn_source_exit_cb(GPid pid, gint status, gpointer source)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+
+	if (sp->exit_cb)
+		sp->exit_cb(status, sp->exit_data);
+
+	g_spawn_close_pid(pid);
+	g_source_unref((GSource *) source);
+}
+
+
+static gboolean spawn_source_prepare(G_GNUC_UNUSED GSource *source, gint *timeout)
+{
+	*timeout = -1;
+	return FALSE;
+}
+
+
+#ifdef G_OS_WIN32
+static gboolean spawn_source_peek_pipe(GPollFD *poll)
+{
+	DWORD available;
+
+	return poll->fd != -1 && (!PeekNamedPipe((HANDLE) _get_osfhandle(poll->fd), NULL, 0, NULL,
+		&available, NULL) || available);
+}
+
+
+static gboolean spawn_source_check(GSource *source)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+
+	return (sp->std[IN].fd != -1 && GetTickCount() - sp->last_send_ticks >= 50) ||

```

Comment:
well, as I read it instead of checking whether it is possible to write to the child stdin it tries every 50ms while the fd is open (it gets closed when everything has been written).  it's a bit odd, but possibly not *so* bad as the stdin write tries to write as much as possible each time, so it "only" waits at most 50ms before trying again.  IIUC